src/librustc_mir/transform/check_unsafety.rs

```diff
@@ -260,45 +276,73 @@ impl<'a, 'tcx> Visitor<'tcx> for UnsafetyChecker<'a, 'tcx> {
-                }
-            }
-        }
-        let within_unsafe = match self.source_scope_local_data[self.source_info.scope].safety {
-            Safety::Safe => {
+        let safety = self.source_scope_local_data[self.source_info.scope].safety;
+        let within_unsafe = match (safety, self.min_const_fn) {
+            // Erring on the safe side, pun intended
+            (Safety::BuiltinUnsafe, true) |
+            // mir building encodes const fn bodies as safe, even for `const unsafe fn`
+            (Safety::FnUnsafe, true) => bug!("const unsafe fn body treated as inherently unsafe"),
+            // `unsafe` blocks are required in safe code
+            (Safety::Safe, _) => {
                 for violation in violations {
-                    if !self.violations.contains(violation) {
-                        self.violations.push(violation.clone())
+                    let mut violation = violation.clone();
+                    if self.min_const_fn {
+                        // overwrite unsafety violation in const fn with a single hard error kind
+                        violation.kind = UnsafetyViolationKind::MinConstFn;
+                    } else if let UnsafetyViolationKind::MinConstFn = violation.kind {
+                        // outside of const fns we treat `MinConstFn` and `General` the same
+                        violation.kind = UnsafetyViolationKind::General;
+                    }
+                    if !self.violations.contains(&violation) {
+                        self.violations.push(violation)
                     }
                 }
                 false
             }
-            Safety::BuiltinUnsafe | Safety::FnUnsafe => true,
-            Safety::ExplicitUnsafe(node_id) => {
+            // regular `unsafe` function bodies allow unsafe without additional unsafe blocks
+            (Safety::BuiltinUnsafe, false) | (Safety::FnUnsafe, false) => true,
+            (Safety::ExplicitUnsafe(node_id), _) => {
+                // mark unsafe block as used if there are any unsafe operations inside
                 if !violations.is_empty() {
                     self.used_unsafe.insert(node_id);
                 }
+                // only some unsafety is allowed in const fn
+                if self.min_const_fn {
+                    for violation in violations {
+                        match violation.kind {
+                            // these are allowed
+                            UnsafetyViolationKind::MinConstFn
+                                // if `#![feature(min_const_unsafe_fn)]` is active
+                                if self.tcx.sess.features_untracked().min_const_unsafe_fn => {},
+                            _ => {
+                                let mut violation = violation.clone();
+                                // overwrite unsafety violation in const fn with a hard error
+                                violation.kind = UnsafetyViolationKind::MinConstFn;
```

Comment:
So first `MinConstFn` is a signal for what is allowed, and now it is used to get a hard error? This is confusing.

Reply:
This is how the other variants are used, too. There's no way around that except to start duplicating a lot of code.

Reply:
In which sense are they also used this way?

Why cannot we have `GeneralAndAllowedInMinConstFn` (strawman proposal :D ) as a new enum variant for, well, previously classified "general" violations that we do permit in `MinConstFn`?

Currently, how would you describe the interpretation of `UnsafetyViolationKind::MinContFn` if you had to write a doc comment for it? I think it'd be something like "either an operation outside of an `unsafe` block that requires `unsafe`, but that we *do* allow for min const; or an `unsafe` operation (maybe in an `unsafe` block) inside a min const fn that is not allowed there". These are two very different things. Which other things are similarly conflated in this check?

Reply:
I was down this rabbit hole yesterday and not able to find a clean solution.

I had multiple enum variants, that ended up with a bunch of `bug!` calls spread around.

I had separate enums, requiring loads of code duplication, because of closures/generators dumping a list of violations, means I'd need to convert that list back to a list of requirements. The unsafety checker as it is (and was before I ever touched it) used the violation struct both as requirements and violations.

Reply:
Well I am afraid the way it works right now will take more time to review than I have this week-end. I'd have to disentangle this spaghetti mess first to understand how *anything* works.

Not good to have the *unsafety checker* out of all passes be so messy :(

Reply:
Note that this enum was created as a hack to support the future incompat lints for extern types and references to fields of packed structs.

This all will clean up nicely when we make those lints a hard error and unify the unsafe logic between `const fn` and `fn`.

Reply:
> This all will clean up nicely when we make those lints a hard error and unify the unsafe logic between const fn and fn.

That is good.